File: system-base/src/main/java/com/levnovikov/system_base/NodeHolder.java
package com.levnovikov.system_base;

import android.util.Log;

import javax.inject.Inject;

/**
 * Author: lev.novikov
 * Date: 17/12/17.
 *
 * NodeHolder
 */

public abstract class NodeHolder<R extends Router> {

    @Inject
    public R router;

    public abstract R build();

    public void destroy() {
        Log.i(">>>>", "destroy " + this.getClass().getSimpleName());
        if (router != null) {
            router.destroyNode();
            router.removeFromBackStack();
            router = null;
        }
    }

    public boolean isActive() {
        return router != null;
    }
}

File: system-base/src/main/java/com/levnovikov/system_base/NodeHolder.kt
package com.levnovikov.system_base

import android.util.Log

import javax.inject.Inject

/**
 * Author: lev.novikov
 * Date: 17/12/17.
 *
 * NodeHolder
 */

abstract class NodeHolder<R : Router> {

    @JvmField
    @Inject
    internal var router: R? = null

    internal fun isActive(): Boolean = router != null

    abstract fun build(): R

    open fun destroy() {
        Log.i(">>>>", "destroy " + this.javaClass.simpleName)
        router?.run {
            detachAllChildren()
            removeFromBackStack()
            this.setBackHandler(null)
            router = null
        }
    }
}
